### DISCUSSION TITLE
Parallel session queries with transactions

### DISCUSSION THREAD

**Discussion by shurshilov** [sqlalchemy/sqlalchemy#10573] [Usage Questions] [answered]
Hi, I'm trying to compare my orm and yours so that I can switch to alchemy. But I have a question about working with transactions. How will the following queries be executed inside alchemy according to the first or second scheme? it's just that both are implemented in my arm. Tell me please
![image](https://github.com/sqlalchemy/sqlalchemy/assets/11828278/8a6805d7-141f-40ae-874b-331b2df42001)

P.S.
in the second scheme, I do not close connections until the entire batch of requests is completed. Thus, in essence, it is like using many connections for a short time versus using one connection but for a long time

**Comment by zzzeek** [comment, marked as answer]
the ORM has a model based around an object called [Session](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-basics) which runs one transaction at a time on a single connection.

There are variants of this `Session` such as the [horizontal sharding](https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.sharding) API which can split a series of unit of work tasks among multiple database backends, which is not really what you seem to have, but it's an example of how custom `Session` implementations can change things around.

The `Session` also allows for custom "binds" that you can split up on models.   So the "multiplex transaction" thing you might be able to approximate... seems like it needs some hacks to work though:


```

binds = {cls: engine.execution_options().connect() for cls in [Email, EmailRecipient, EmailLog, EmailAttribute]}
session = Session(
    binds=binds
)
```

I'm not really sure how this idea "works" because you can't do things like have foreign keys between the tables, unless you set the isolation level down to read uncommitted which would be extremely unusual.  Also you lose durability if you aren't using something like 2pc.

Here's a demo with PG, but I can't use any foreign keys


```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import foreign
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]
    bs: Mapped[list[B]] = relationship(
        "B", primaryjoin=lambda: A.id == foreign(B.a_id)
    )


class B(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)
    a_id: Mapped[int] = mapped_column()
    data: Mapped[str]


e = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.drop_all(e)
Base.metadata.create_all(e)

c1, c2 = (
    e.execution_options(logging_token="A").connect(),
    e.execution_options(logging_token="B").connect(),
)
s = Session(binds={A: c1, B: c2})

s.add(A(data="a1", bs=[B(data="b1")]))
s.commit()

c1.close()
c2.close()

```

output:

```
2023-11-01 12:59:25,430 INFO sqlalchemy.engine.Engine [A] BEGIN (implicit)
2023-11-01 12:59:25,431 INFO sqlalchemy.engine.Engine [A] INSERT INTO a (data) VALUES (%(data)s) RETURNING a.id
2023-11-01 12:59:25,432 INFO sqlalchemy.engine.Engine [A] [generated in 0.00019s] {'data': 'a1'}
2023-11-01 12:59:25,433 INFO sqlalchemy.engine.Engine [B] BEGIN (implicit)
2023-11-01 12:59:25,434 INFO sqlalchemy.engine.Engine [B] INSERT INTO b (a_id, data) VALUES (%(a_id)s, %(data)s) RETURNING b.id
2023-11-01 12:59:25,434 INFO sqlalchemy.engine.Engine [B] [generated in 0.00016s] {'a_id': 1, 'data': 'b1'}
2023-11-01 12:59:25,436 INFO sqlalchemy.engine.Engine [B] COMMIT
2023-11-01 12:59:25,448 INFO sqlalchemy.engine.Engine [A] COMMIT

```

**Reply by shurshilov** [reply]
I think you could try to do this with PREPARE TRANSACTION instead "read uncommitted"?
P.S. in postgres "read uncommitted" = "read committed", but i fully understood you

**Reply by zzzeek** [reply]
I dont see how PREPARE TRANSACTION helps with what's described.

**Reply by shurshilov** [reply]
yes, I understood, I can do the same as in my orm, the main thing is that there are no FK keys)
My orm also not "work" if exist FK.
thank you!
But it turns out in this case I lose the cascade delete functionality?
